tests/sentry/mail/activity/test_release_summary.py

```diff
@@ -52,6 +54,9 @@ def test_simple(self):
                     data={"version": self.release.version, "deploy_id": self.deploy.id},
                 )
             )
+            release_summary = create_notification_with_properties(
```

Comment:
Why is it that this needs to be redefined in the test rather than this happening when the `ReleaseSummaryActivityNotification` is created?

Reply:
@ceorourke This test is a little unique, we don't usually create notifications and test them without a provider. In the actual flow notifications are always sent to a provider, so we populate these fields there in the message builder.

- I agree that this solution of adding properties to the object after it is created is not great but the notification hierarchy is supposed to be independent of the provider, but it does expect to know the `provider` and `url_format` to create the notification. 
- I couldn't create different types of notification using inheritance because the notification object is already created by the time we know if it is a Slack or MS Teams notification.
- Another option was to pass in these parameters to functions wherever they're required, but that would be a little more uglier in my opinion.

Let me if you have any other suggestions.